UI: Apply user updates for mobile navigation accessibility

yatishgoel opened the pull request:
Closes #31380

<!-- If your PR is related to an issue, provide the number(s) above; if it resolves multiple issues, be sure to break them up (e.g. "closes #1000, closes #1001"). -->

<!--

Thank you for contributing to Storybook! Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/contribute

-->

## What I did

This PR significantly enhances the accessibility of the mobile navigation and addresses focus management issues, resolving the core concerns raised in #31380. It also incorporates the DOM order fix for desktop and mobile reading consistency from #31076.

**Key improvements:**

1.  **ARIA Attributes for Mobile Navigation:**
    * The mobile menu toggle button now correctly uses `aria-label`, `aria-expanded`, and `aria-controls` to properly communicate its state and relationship with the menu drawer (`MobileNavigation.tsx`).
    * The addon panel toggle button also includes `aria-label` and `aria-expanded`.

2.  **Accessible Mobile Menu Drawer (`MobileMenuDrawer.tsx`):**
    * The menu drawer now has `role="dialog"`, `aria-modal="true"`, and an `aria-label` for clear anouncement to screen reader users.
    * The content area within the drawer (holding the navigation links) is marked with `role="navigation"`.

3.  **Focus Management and Trapping:**
    * A robust `useFocusTrap` hook has been implemented and applied to `MobileMenuDrawer.tsx`.
    * When the mobile menu opens, focus is now programmatically moved into the drawer, targeting the first focusable element.
    * Focus is trapped within the open menu, allowing navigation via `Tab` and `Shift+Tab`, and closing via the `Escape` key.
    * Focus is correctly restored to the previously focused element (the menu button) when the drawer is closed.
    * The main content area is marked with `aria-hidden="true"` and `tabIndex="-1"` when the mobile menu is open to prevent interaction with background content (`Layout.tsx`).

4.  **DOM Order for Reading Consistency (Addresses #31076):**
    * **Desktop View:** The DOM order in `Layout.tsx` ensures `<SidebarContainer>` renders before the main content (`<MainContentMatcher>`) for a logical reading order.
    * **Mobile View:** The `<MobileNavigation>` trigger component is rendered first in the DOM within its mobile layout block, while CSS `order` is used for its visual placement at the bottom. This ensures screen readers encounter the navigation trigger in a logical sequence.

5.  **General Code Refinements:**
    * Centralized initial focus logic within `useFocusTrap`.
    * Improved focusable element detection in `useFocusTrap`.
    * Added `id` props and `aria-controls` for the `MobileAddonsDrawer` as well, preparing it for similar focus trap implementation if it contains interactive elements.


## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_
_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

1.  Run a sandbox for a relevant template, e.g., `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2.  Open Storybook in your browser.
3.  **Desktop View:**
    * Using a screen reader (e.g., NVDA, VoiceOver), verify that the sidebar content is announced before the main story/docs content.
4.  **Mobile View (resize browser or use dev tools):**
    * Verify the mobile navigation bar appears visually at the bottom.
    * Using a screen reader:
        * Check that the "Open navigation menu" button is focusable and correctly announces its purpose and state (collapsed).
        * Activate the button.
        * Verify the "Navigation menu" dialog is announced as open and modal.
        * Verify focus moves into the menu (e.g., to the first navigation item).
        * Navigate within the menu using `Tab` and `Shift+Tab`. Ensure focus stays within the menu.
        * Press `Escape`. Verify the menu closes and focus returns to the "Open navigation menu" button.
        * Verify the main content behind the menu was hidden from the screen reader (`aria-hidden="true"`) when the menu was open.
    * Repeat similar checks for the "Open addon panel" button and its drawer if it contains interactive elements (note: focus trap for addon drawer is suggested but might be implemented separately if not part of this direct change).


<!-- Please include the steps to test your changes here. For example:

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Open Storybook in your browser
3. Access X story

-->

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

<!-- BENCHMARK_SECTION -->
<!-- BENCHMARK_SECTION -->


<!-- greptile_comment -->

## Greptile Summary

Here's a concise summary of the key accessibility improvements made to the mobile navigation components:

Enhanced mobile navigation components with comprehensive accessibility features and proper DOM ordering.

- Added proper ARIA attributes to navigation elements (`role="dialog"`, `aria-modal`, `aria-controls`, `aria-expanded`) in `MobileNavigation.tsx` and `MobileMenuDrawer.tsx`
- Implemented focus trap management with new `useFocusTrap` hook for modal dialogs
- Fixed DOM order in `Layout.tsx` to ensure sidebar renders before main content for logical screen reader flow
- Added `aria-hidden` and `tabIndex` controls to main content when mobile menu is open
- Improved keyboard navigation with proper focus restoration and trap management within modal dialogs

The changes significantly improve screen reader compatibility and keyboard navigation while maintaining visual layout through CSS ordering rather than DOM manipulation.



<!-- /greptile_comment -->

yatishgoel committed:
Apply user updates for mobile navigation accessibility

yatishgoel committed:
Refactor useFocusTrap hook for improved clarity and performance by removing unnecessary comments and simplifying focus handling logic.

yatishgoel committed:
Enhance type safety in MobileMenuDrawer and useFocusTrap; streamline aria-controls in MobileNavigation

yatishgoel committed:
Merge branch 'next' into fix/mobile-nav-a11y-user-updates

yatishgoel committed:
Removed duplicate MainContentMatcher and MainContentMatcher

ndelangen committed:
Merge branch 'next' into fix/mobile-nav-a11y-user-updates

ndelangen commented:
@MichaelArestad Do you have any input here?

MichaelArestad commented on code/core/src/manager/components/layout/Layout.tsx:
```diff
@@ -126,10 +126,21 @@ const useLayoutSyncingState = ({
+        <ContentContainer
+          shown={!!match}
+          aria-hidden={isMobileMenuOpen}
+          tabIndex={isMobileMenuOpen ? -1 : undefined}
```
Is there a way to do this without using a negative tabIndex?

yatishgoel replied:
Hi @MichaelArestad,

Thanks for the feedback!

Regarding the `tabIndex={-1}` on the `ContentContainer` in `Layout.tsx`:

I used `tabIndex={isMobileMenuOpen ? -1 : undefined}` in conjunction with `aria-hidden={isMobileMenuOpen}` as a common pattern to ensure that when the mobile menu is open, the main page content is both hidden from assistive technologies (via `aria-hidden`) and removed from the keyboard tabbing order (via `tabIndex={-1}`). This helps create a "modal" experience, preventing users from unintentionally interacting with the underlying page content while the menu is active.

The goal was to make the background content effectively inert so that focus is correctly trapped within the mobile menu.

You're right to question if there are other ways to achieve this. The `inert` attribute is indeed a more modern and semantically appropriate solution for this exact use case, as it handles both the focus and accessibility tree aspects.

Given its strong browser support now, I'd be happy to update the `ContentContainer` to use `inert={isMobileMenuOpen}` instead of managing `tabIndex` and `aria-hidden` separately if that's preferred. This would simplify the logic and rely on a more direct HTML feature.

Let me know your thoughts!

MichaelArestad replied:
@yatishgoel Thank you for the explanation. I dug a bit deeper and I think you made the right call there. Much appreciated!

yatishgoel replied:
Hi @Sidnioulz,

Thanks for your feedback on using `inert` or `showModal()` for the mobile navigation. I'm planning to refactor the mobile menu and addons panel based on your suggestion.

The idea is to:
1.  Replace the current `div` elements (that have `role="dialog"`) with actual HTML `<dialog>` elements.
2.  Use `dialog.showModal()` to open them and `dialog.close()` to close them. This should provide native focus trapping and make the rest of the page `inert` automatically.
3.  This would allow us to remove the custom focus trap hook and the manual `aria-hidden`/`tabindex` attributes previously applied to the main content area.

Does this approach align with what you had in mind for a more robust and clearer implementation?

Thanks!

MichaelArestad reviewed (changes requested):
@yatishgoel Nice!

@ndelangen These changes seem great. I am not an expert on aria attributes and I have not implemented anything like `useFocusTrap` though I understand the reasoning.

I believe this breaks a few Interaction tests in Navigation that were targeting elements with a specific `title` attribute. We shouldn't be using `title` attributes for this anyway. `title` attributes are not designed to be used for anything other than iframes.

Regarding the `title` attributes, they were being used as tooltips, which isn't great. In the near future, I do think we should probably:
1. Create new tooltips that are accessible, specifically the tooltips designed to show a short string (black background, white text).
2. Implement those tooltips in almost every situation where we use `title`.

MichaelArestad commented:
@ndelangen Can you code review this when you get a chance? Thank you!

Sidnioulz commented on code/core/src/manager/components/mobile/navigation/useFocusTrap.tsx:
```diff
@@ -0,0 +1,122 @@
+      const focusableElements = getFocusableElements(container);
+
+      // If no focusable elements and not Escape, prevent default and exit
+      if (focusableElements.length === 0 && e.key !== 'Escape') {
+        e.preventDefault();
+        return;
+      }
```
I wonder if that would break built-in keyboard shortcuts in screenreaders, e.g. those used to navigate between headings or paragraphs.

Sidnioulz commented on code/core/src/manager/components/mobile/navigation/useFocusTrap.tsx:
```diff
@@ -0,0 +1,122 @@
+
+      // Proceed with Tab logic only if there are focusable elements
+      if (focusableElements.length > 0 && e.key === 'Tab') {
+        const firstElement = focusableElements[0];
+        const lastElement = focusableElements[focusableElements.length - 1];
+
+        if (e.shiftKey) {
+          if (document.activeElement === firstElement) {
+            e.preventDefault();
+            lastElement.focus();
+          }
+        } else {
+          if (document.activeElement === lastElement) {
+            e.preventDefault();
+            firstElement.focus();
+          }
+        }
+      }
```
You don't account for DOM mutations inside the dialog.

Sidnioulz commented on code/core/src/manager/components/mobile/navigation/useFocusTrap.tsx:
```diff
@@ -0,0 +1,122 @@
+        }
+      });
+
+      const focusableElements = getFocusableElements(container);
```
Why is that a different item than the one before? I am not understanding the distinction.

Sidnioulz reviewed (changes requested):
Thanks so much for undertaking this work, @yatishgoel! This is hard work so please don't take anything in my review as a criticism of your effort. I very much appreciate you giving this a go!

You applied a `dialog` `role` with an `aria-modal` attribute. That's great, and it's what I wrote in the issue so really this is my own fault! It's unfortunately missing a powerful element of the Dialog element: `showModal`. The `showModal` function lets you instruct the browser to open a modal and install a focus trap for you, which greatly reduces the amount of code to write and makes better use of standards.

You also used the `navigation` `role` in your code. It would be preferrable to use the `nav` element directly instead. Generally speaking (and as reminded on https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Roles/navigation_role), it's preferable to use the HTML standard element rather than its inferred role and attributes.

I shouldn't have mentioned `role` in my issue. I sometimes refer to roles instead of elements out of habit because I don't know if the HTML markup is controlled and can be edited, but when a role can be fulfilled through the use of an element, it should. That allows for better future compatibility and easier maintenance.

To move forward, I suggest removing the focus trap implementation and using `showModal` instead, and switching to actual nav/dialog elements. Then, there would only be minor details left to address.

yatishgoel committed:
Refactor mobile navigation components to improve accessibility and state management. Updated MobileAddonsDrawer and MobileMenuDrawer to use dialog elements for better focus control. Removed unused useFocusTrap hook. Simplified MainContentMatcher by removing unnecessary props. Enhanced MobileNavigation to manage addon panel state more effectively.

yatishgoel committed:
Merge branch 'next' into fix/mobile-nav-a11y-user-updates

Sidnioulz commented on code/core/src/manager/components/mobile/navigation/MobileAddonsDrawer.tsx:
```diff
@@ -1,22 +1,69 @@
+  isOpen,
+  onClose,
+}) => {
+  const dialogRef = useRef<HTMLDialogElement>(null);
```
I feel you could move much of this in a hook in `code/core/src/manager/hooks/`. That would remove duplicate code in your PR and greatly help in reapplying this code for other popups.

Sidnioulz replied:
Just FYI, in case you're curious, and definitely not part of this PR: There is a `code/core/src/components/components/Modal/Modal.tsx` component that could be simplified by using your hook and native HTML elements. It still has a distinct API so I don't think it should be edited here, but should you be interested in making further contributions, we can have a chat about it.

yatishgoel replied:
Hey @Sidnioulz—thanks for pointing that out! I’d definitely be interested in refactoring `code/core/src/components/components/Modal/Modal.tsx` with the new hook.
Can we open a separate issue for it so we can track the work and chat there (or on Discord) about the scope and next steps?

Sidnioulz commented on code/core/src/manager/components/mobile/navigation/MobileAddonsDrawer.tsx:
```diff
@@ -1,22 +1,69 @@
+    const dialogNode = dialogRef.current;
+    if (dialogNode) {
+      if (isOpen) {
+        if (!dialogNode.hasAttribute('open')) {
+          dialogNode.showModal();
+        }
+      } else {
+        if (dialogNode.hasAttribute('open')) {
+          dialogNode.close();
+        }
```
Works great overall, though for some reason pressing <kbd>Esc</kbd> does not close the modal. The keyboard event must be consumed by someone else.

Could you please investigate where it is consumed? If you can't find a  source, maybe the maintainers will be ok with reusing Radix's close-on-escape utility which is already a dependency. I've asked whether it's ok to reuse Radix or not on Discord (#contributing).

yatishgoel replied:
Hey @Sidnioulz, turns out our global shortcut listener was calling `preventDefault()` on Escape before `<dialog>` could handle it. I added a quick check to skip shortcuts when the event is inside an open `<dialog>` (and inputs), so now Escape closes the modal as expected. Does this minimal tweak look good, or should we pull in Radix’s close-on-escape instead?

Sidnioulz reviewed (commented):
A small bug left to fix and a refactor could be made to regroup the modal hooks into a single file, but this is great overall and would definitely fix the linked issue. Kudos, @yatishgoel!

yatishgoel committed:
Refactor MobileAddonsDrawer and MobileMenuDrawer to utilize useModalDialog hook for improved dialog management. Introduced useModalDialog for better encapsulation of dialog logic and enhanced accessibility. Updated shortcut handling to skip events when a dialog is open.

yatishgoel committed:
Merge branch 'next' into fix/mobile-nav-a11y-user-updates

Sidnioulz committed:
Merge branch 'next' into fix/mobile-nav-a11y-user-updates

Sidnioulz commented on code/core/src/manager-api/modules/shortcuts.ts:
```diff
@@ -149,9 +149,17 @@ export const defaultShortcuts: API_Shortcuts = Object.freeze({
+function shouldSkipShortcut(event: KeyboardEvent) {
   const target = event.target as Element;
-  return /input|textarea/i.test(target.tagName) || target.getAttribute('contenteditable') !== null;
+  if (/input|textarea/i.test(target.tagName) || target.getAttribute('contenteditable') !== null) {
+    return true;
+  }
+  const dialogElement = target.closest('dialog[open]');
+  if (dialogElement) {
+    return true;
+  }
+
+  return false;
```
@MichaelArestad @ghengeveld what do you think should happen here?

The goal is to make sure the global shortcut manager doesn't consume `Esc` key presses when a HTML `dialog` is open through `showModal`. Currently, the shortcut manager consumes `Esc` and so prevents closing a modal HTML dialog.

Things that need to be checked by people who know the codebase better than me:
* Is it ok to remove *all* keyboard shortcuts when a modal is open, or do some still make sense? Personally, I'd say it's ok because we don't want the user to accidentally break out of the modal and virtually all keys being handled trigger focus or navigation changes
* What should happen to the `Esc` handlers in that file, ultimately? Shouldn't we remove the bit about mobile navigation in this very PR since it's no longer relevant once the mobile nav has been rewritten as a `dialog`? Or am I missing something?

Sidnioulz commented on code/core/src/manager-api/modules/shortcuts.ts:
```diff
@@ -390,7 +398,7 @@ export const init: ModuleFn = ({ store, fullAPI, provider }) => {
     // Listen for keydown events in the manager
     document.addEventListener('keydown', (event: KeyboardEvent) => {
-      if (!focusInInput(event)) {
+      if (!shouldSkipShortcut(event)) {
```
This worked for the addon drawer, showing this specific code path no longer disturbs HTML dialogs.

However, I still can't close the mobile nav dialog with `Esc`, something else is blocking it :(

yatishgoel replied:
> This worked for the addon drawer, showing this specific code path no longer disturbs HTML dialogs.
> 
> However, I still can't close the mobile nav dialog with `Esc`, something else is blocking it :(

- In `useModalDialog`, I added a `keydown` event listener specifically for the Esc key
- When Esc is pressed, it calls `preventDefault()` and `onClose()` instead of immediately closing
- This ensures the Esc key triggers our state-managed close sequence (with animation) rather than the browser's immediate close

**Before:** Mobile menu wasn't a proper dialog → Shortcut manager consumed Esc → Dialog couldn't close
**After:** Proper dialog element → Shortcut manager skips → Dialog handles Esc → Triggers animated close sequence

Sidnioulz commented on code/core/src/manager/components/mobile/navigation/MobileMenuDrawer.tsx:
```diff
@@ -1,44 +1,59 @@
       <Transition
-        nodeRef={containerRef}
+        nodeRef={dialogRef}
         in={isMobileMenuOpen}
         timeout={MOBILE_TRANSITION_DURATION}
         mountOnEnter
         unmountOnExit
-        onExited={() => setMobileAboutOpen(false)}
+        onExited={() => {
+          setMobileAboutOpen(false);
+          if (dialogRef.current && dialogRef.current.hasAttribute('open')) {
+            dialogRef.current.close();
+          }
+        }}
       >
```
Not sure what's going on here, but the exit animation is broken nonetheless. I need to brush up on React transitions to be of help here, but the `close()` call from native dialog is immediate, so it probably should happen at the end of the exit transition, with another layer of state management to handle triggering the exit transition.

Let's see what the people who have written React code in the past 3 years say, because I'm very rusty :D

Sidnioulz replied:
This seems to be fixed now, resolving.

Sidnioulz commented on code/core/src/manager/hooks/useModalDialog.ts:
```diff
@@ -0,0 +1,43 @@
+          dialogNode.showModal();
+        }
+      } else {
+        if (dialogNode.hasAttribute('open')) {
+          dialogNode.close();
+        }
```
This might be where you need to do something to trigger unmounting the Transition and letting the Transition do the closing for you. Which ironically might mean we need to preserve the `Esc` handler for this use case :( 

We need to go through this mess because Firefox does not yet support `allow-discrete` that would let us use native CSS to animate the exit animation of the dialog (see https://frontendmasters.com/blog/the-dialog-element-with-entry-and-exit-animations/ and https://codepen.io/sidnioulz/pen/PwqNQzE for a simple relevant codepen).

@ghengeveld your thoughts here on how to proceed would be appreciated.

yatishgoel committed:
Refactor MobileMenuDrawer and useModalDialog for improved dialog handling

- Introduced `forceCloseDialog` function in MobileMenuDrawer to encapsulate dialog closing logic.
- Updated onExited callback to use `forceCloseDialog` for cleaner code.
- Enhanced useModalDialog to handle Escape key for closing the dialog and prevent default behavior.
- Removed redundant dialog close logic from useModalDialog to streamline the component's functionality.

yatishgoel commented on code/core/src/manager/components/mobile/navigation/MobileMenuDrawer.tsx:
```diff
@@ -1,44 +1,59 @@
             >
               {(sidebarState) => (
-                <SidebarContainer ref={sidebarRef} state={sidebarState}>
+                <SidebarContainer ref={sidebarRef} state={sidebarState} role="navigation">
```
Thanks for pointing this out. I have updated this part.

yatishgoel committed:
Merge branch 'next' into fix/mobile-nav-a11y-user-updates

Sidnioulz commented on code/core/src/manager/hooks/useModalDialog.ts:
```diff
@@ -0,0 +1,43 @@
+      };
+      dialogNode.addEventListener('close', handleDialogCloseEvent);
+      return () => {
+        dialogNode.removeEventListener('close', handleDialogCloseEvent);
```
Thanks @yatishgoel! I'll test this with a screenreader soon and we can hopefully get this in :) Thanks for your continued contribution!

Sidnioulz reviewed (changes requested):
Congrats on fixing the animations! Can you please test the mobile addon panel again? I see a few remaining issues:
* It's positioned in the middle of the screen instead of the bottom
* The close button doesn't close the panel
* It seems I have to press Esc 2-3 times to close the panel

yatishgoel committed:
fix(mobile): Fix dialog positioning and close behavior

- Fix MobileMenuDrawer and MobileAddonsDrawer appearing centered instead of bottom
- Add full width positioning for mobile dialogs
- Fix Esc key requiring double press in addon drawer
- Fix close button not working properly
- Ensure smooth exit animations with proper timing coordination
- Override browser default modal positioning with &[open] selector

yatishgoel commented:
> Congrats on fixing the animations! Can you please test the mobile addon panel again? I see a few remaining issues:
> 
> * It's positioned in the middle of the screen instead of the bottom
> * The close button doesn't close the panel
> * It seems I have to press Esc 2-3 times to close the panel


Thanks for the feedback! I actually just fixed those exact issues in both MobileAddonsDrawer.tsx and MobileMenuDrawer.tsx:

**Positioning:** Added `position: 'fixed', bottom: 0, right: 0, left: 0, top: 'auto'` with `&[open]` selector to override browser default modal positioning, plus `maxWidth: '100vw'` for full width.

**Close Behavior:** Converted MobileAddonsDrawer to use React Transition with the same working pattern as MobileMenuDrawer - added `forceCloseDialog` function that runs after exit animation completes. This should fix both the close button and single Esc key press.

Both mobile drawers now use identical patterns: React Transition + useModalDialog + forceCloseDialog for behavior, and the same CSS positioning overrides. MobileMenuDrawer was working correctly, so I used it as the reference to fix MobileAddonsDrawer. Both should now appear at bottom with full width and proper close behavior.

yatishgoel committed:
Merge branch 'next' into fix/mobile-nav-a11y-user-updates

Sidnioulz reviewed (approved):
Code does everything I'd like it to do! Now this is just a matter of getting the CI passing.

We can now turn our attention to the CI :) I haven't looked in detail at the CI reports, but if you get stuck, feel free to ask me to take over and I will have a look. It seems there are a few tests that need to be updated, maybe you could start with this? https://app.circleci.com/jobs/github/storybookjs/storybook/843273

Sidnioulz commented:
@yatishgoel can you please update the tests affected by your change?

yatishgoel committed:
Merge branch 'next' into fix/mobile-nav-a11y-user-updates

yatishgoel committed:
fix: update mobile navigation tests to use aria-label instead of title

- Replace getByTitle() with getByLabelText() in test queries
- Update MockPanel component to use aria-label instead of title
- Align tests with accessibility improvements that replaced title with aria-label attributes
- All MobileNavigation.stories.tsx tests now pass

yatishgoel committed:
fix: update e2e composition test to use aria-label selector

- Replace 'button[title="Open navigation menu"]' with 'button[aria-label="Open navigation menu"]'
- Fixes timeout issue in composition.spec.ts mobile test
- Aligns e2e test with accessibility improvements that replaced title with aria-label

yatishgoel committed:
fix: update e2e tests to use aria-label for navigation elements

- Replace title attributes with aria-label in manager.spec.ts for better accessibility
- Aligns with recent accessibility improvements in the codebase

yatishgoel committed:
fix: add aria-label to mobile addon panel close button

- Replace title with aria-label for mobile close button in Panel.tsx
- Fixes e2e test failure expecting 'aria-label="Close addon panel"'
- Completes accessibility improvements for mobile addon panel
- Aligns mobile panel close button with navigation accessibility standards

yatishgoel commented:
> @yatishgoel can you please update the tests affected by your change?


Hi @Sidnioulz,

I have successfully updated all tests affected by the accessibility changes and confirmed they're working locally.

- **`MobileNavigation.stories.tsx`** - Updated selectors from `getByTitle()` to `getByLabelText()` (8/8 tests passing)
- **`e2e-tests/manager.spec.ts`** - Updated mobile navigation selectors to use `aria-label` attributes  
- **`composition.spec.ts`** - Fixed mobile navigation selector
- **`Panel.tsx`** - Added missing `aria-label="Close addon panel"` for mobile close button
- **Local testing**: All 30 related tests pass, confirming accessibility fixes are production-ready

Thanks for your guidance!
